Properly display owned artifacts by querying distgit

ryanlerch opened the pull request:
Signed-off-by: Ryan Lerch <rlerch@redhat.com>

ryanlerch committed:
Query distgit/pagure for artifacts owned by users/groups

Signed-off-by: Ryan Lerch <rlerch@redhat.com>

abompard commented:
Could you rebase and fix the conflict please?

ryanlerch commented:
> Could you rebase and fix the conflict please?

done.

not sure what the failing test is from, but its happening in the main branch too

abompard commented on fmn/api/distgit.py:
```diff
@@ -25,6 +26,37 @@ async def get_projects(self, pattern):
+            firstpage.raise_for_status()
+            _appendArtifacts(firstpage.json()["projects"])
+
+            if firstpage.json()["pagination"]["pages"] > 1:
```
You could simplify that with a `while` loop. Check out `fmn.rules.services.distgit:DistGitService._all_values` for an example.

ryanlerch replied:
okay... so i implemented this -- and it worked -- however it started to make the tests a lot more difficult
here is the branch with the change, and my attempts at redoing the tests

https://github.com/ryanlerch/fmn/tree/artifacts2-while-fullurl

basically, i previously used the same response twice with return_value, however, getting async coroutine StopIteration issues when trying to use side_effect (returning a list) to do multiple different return values.

is there something like responses for httpx async here to make this easier?

ryanlerch replied:
I think i have ti sorted -- PR changes incoming soon

ryanlerch commented:
Ok! using the "next" url returned by pagure now in a while loop.

Got the tests working now too with pytest-httpx to mock the responses which works a lot better than what i was doing before.

Kept the initial commit in there, but intend to squash them into 1 before merging

abompard commented on fmn/api/distgit.py:
```diff
@@ -25,6 +25,36 @@ async def get_projects(self, pattern):
+            params["page"] = "1"
+            response = await self.client.get(endpoint, params=params)
+            response.raise_for_status()
+            print(response.json())
```
Leftover here :-)

abompard commented on fmn/api/distgit.py:
```diff
@@ -25,6 +25,36 @@ async def get_projects(self, pattern):
+
+            _appendArtifacts(response.json()["projects"])
+            while response.json().get("pagination", {}).get("next"):
+                response = await self.client.get(response.json()["pagination"]["next"])
```
Yeah that works. Another way of doing it would have been to initialize `url` with the initial URL and refresh it with the next URL in the response, such as:
```python
url = f"{endpoint}?{urlencode(params)}"
while url:
    response = await self.client.get(url)
    response.raise_for_status()
    data = response.json()
    _appendArtifacts(data["projects"])
    url = data.get("pagination", {}).get("next")
```
This would allow you to avoid repeating yourself when writing the `client.get` calls, the `raise_for_status`, `_appendArtifacts` and such.

abompard commented on fmn/api/distgit.py:
```diff
@@ -25,6 +25,36 @@ async def get_projects(self, pattern):
         return result["projects"]
 
+    async def get_owned(self, names, user_or_group):
+        def _appendArtifacts(projects):
```
That's javascript naming scheme, please use `_append_artifacts`.

abompard commented on tests/api/test_handlers.py:
```diff
@@ -285,35 +286,51 @@ def test_get_applications(self, client):
+            response = client.get(f"{self.path}/artifacts/owned", params={"groups": ["dudegroup"]})
 
-        assert response.status_code == status.HTTP_200_OK
+        print(httpx_mock.get_requests())
```
Not sure you need that either.

ryanlerch commented:
Thanks for the review!

Have implemented your changes, and squashed the commits now too.

ryanlerch commented:
Okay! Updated the tests to use respx (which was already being used) rather than pytest-httpx (which i introduced in a previous iteration of thius PR before knowing about respx)
